Std/Data/List/Basic.lean

```diff
@@ -889,10 +889,18 @@ sublists [1, 2, 3] = [[], [1], [2], [1, 2], [3], [1, 3], [2, 3], [1, 2, 3]]
 ```
 -/
 def sublists (l : List α) : List (List α) :=
+  l.foldr (fun a acc => acc.bind fun x => [x, a :: x]) [[]]
+
+/-- A version of `List.sublists` that has faster runtime performance but worse kernel performance -/
+def sublistsFast (l : List α) : List (List α) :=
   let f a arr := arr.foldl (init := Array.mkEmpty (arr.size * 2))
     fun r l => (r.push l).push (a :: l)
   (l.foldr f #[[]]).toList
 
+-- The fact that this transformation is safe is proved in mathlib. Using a `csimp` lemma here is
+-- impractical as we are missing a lot of lemmas about lists.
+attribute [implemented_by sublistsFast] sublists
```

Comment:
These lemmas should be upstreamed, then.

Reply:
I think this is a reasonable number of lemmas across multiple files; even if we did upstream them, they would probably need to go in a different file so as not to make this one grow too much, and then anyone importing only this file would still not get the csimp lemma.

Is this you drawing a hard line saying "we don't ever want to use `implemented_by` in this way", or the weaker "we should put upstreaming these results on our to-do list"? Ideally we could merge this before doing any upstreaming, as everyone downstream in mathlib (multiset and Finset power sets) inherits the speed boost.

An alternative that wouldn't have to wait for upstreaming is to define `sublistsFastKernel` in mathlib, and then change all the downstream lemmas and definitions to be about that instead.

Reply:
I think if `sublists` has too many requirements we should just take it out of `Data.List.Basic` and move it to `Data.List.Sublists` with appropriate imports.

Reply:
Ok, so this PR should be considered blocked by upstreaming the required results, and having the `csimp` lemma in `Mathlib` is foo far away?

Reply:
If you think the lemma stuff is going to take a while and don't want that to hold things up here, at least put in a TODO and/or open an issue tracking moving those lemmas to std.